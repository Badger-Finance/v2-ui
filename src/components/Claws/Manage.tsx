import React, { FC, useState } from 'react';
import { Box, Button, Container, Grid, MenuItem, Select } from '@material-ui/core';
import { useMainStyles } from './index';
import ClawParams, { ClawParam } from './ClawParams';
import ClawLabel from './ClawLabel';
import ClawDetails from './ClawDetails';

const depositOptions = ['wBTCwETHSLP', 'bBadger'];
const withdrawOptions = ['eCLAW FEB29', 'eCLAW MARCH29'];

const eCLAWS: Record<string, string> = {
	wBTCwETHSLP: '0.000017',
	bBadger: '0.000017',
	'eCLAW FEB29': '1000',
	'eCLAW MARCH29': '2000',
};

const options: Record<string, string[]> = {
	deposit: depositOptions,
	withdraw: withdrawOptions,
};

const initialValue = { amount: '0.00' };

const Manage: FC = () => {
	const classes = useMainStyles();
	const [mode, setMode] = useState<'deposit' | 'withdraw'>('deposit');
	const [txParams, setTxParams] = useState<ClawParam>(initialValue);

	return (
		<Container>
			<Box pb={1}>
				<Grid item xs={12} sm={3} style={{ margin: 'auto' }}>
					<Select
						displayEmpty
						fullWidth
						variant="outlined"
						color="secondary"
						value={mode}
						style={{ textAlign: 'center' }}
						onChange={(v: any) => {
							setTxParams(initialValue);
							setMode(v.target.value);
						}}
					>
						<MenuItem value="" disabled>
							Select Mode
						</MenuItem>
						<MenuItem value="deposit">DEPOSIT</MenuItem>
						<MenuItem value="withdraw">WITHDRAW</MenuItem>
					</Select>
				</Grid>
			</Box>
			<Grid item xs={12}>
				<Box clone pb={1}>
					<Grid item xs={12}>
						<ClawLabel
							name="Token"
							balanceLabel={txParams.selectedOption ? `Available ${txParams.selectedOption}:` : ''}
							balance={txParams.selectedOption ? eCLAWS[txParams.selectedOption] : ''}
						/>
					</Grid>
				</Box>
				<Grid item xs={12}>
					<ClawParams
						referenceBalance={txParams.selectedOption ? eCLAWS[txParams.selectedOption] : ''}
						placeholder="Select Token"
						amount={txParams.amount}
						onAmountChange={(amount: string, error?: boolean) => {
							setTxParams({
								...txParams,
								amount,
								error: error ? `Amount exceeds ${txParams.selectedOption} balance` : undefined,
							});
						}}
						selectedOption={txParams.selectedOption}
						options={options[mode]}
						onOptionChange={(selectedOption: string) => {
							setTxParams({
								selectedOption,
								amount: '0.00',
								error: undefined,
							});
						}}
						disabledAmount={!txParams.selectedOption}
					/>
				</Grid>
				<Grid item xs={12}>
					<Grid container className={classes.details}>
						<ClawDetails
							details={[
								{ 'Liquidation Price': '1.000' },
								{ 'Collateral Ratio - Global': '1.2x' },
								{ 'Collateral Ratio - Minimum': '1.2x' },
								{ 'Collateral Ratio - Current': '4x' },
								{ 'Expiration 4x': 'Feb 29th, 2021' },
								{ 'Minimum Deposit': '100 eCLAW' },
							]}
						/>
					</Grid>
				</Grid>
				<Grid xs={12}>
					<Grid container>
						<Button
							color="primary"
							variant="contained"
							disabled={!!txParams.error || !txParams.selectedOption}
							size="large"
							className={classes.button}
						>
							{txParams.error ? txParams.error : mode.toLocaleUpperCase()}
						</Button>
					</Grid>
				</Grid>
			</Grid>
		</Container>
	);
};

export default Manage;
