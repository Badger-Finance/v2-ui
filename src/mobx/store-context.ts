import { RootStore } from './store';
import { createContext } from 'react';

// Use React context to make your store available in your application
export const StoreContext = createContext({} as RootStore);
export const StoreProvider = StoreContext.Provider;
