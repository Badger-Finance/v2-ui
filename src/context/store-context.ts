import { createContext } from "react";
import { RootStore } from "../mobx/store";

// Use React context to make your store available in your application
export const StoreContext = createContext({} as RootStore);
export const StoreProvider = StoreContext.Provider;