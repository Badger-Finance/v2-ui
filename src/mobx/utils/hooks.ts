import React from 'react';
import { StoreContext } from 'mobx/store-context';
import BigNumber from 'bignumber.js';

export function useConnectWallet() {
	const store = React.useContext(StoreContext);
	const { onboard } = store.wallet;

	return async () => {
		if (store.uiState.sidebarOpen) store.uiState.closeSidebar();

		const walletSelected = await onboard.walletSelect();
		const walletReady = await onboard.walletCheck();

		if (walletSelected && walletReady) {
			store.wallet.connect(onboard);
		}
	};
}

export function useBdiggToDigg() {
	const store = React.useContext(StoreContext);
	const { stats, rebaseStats } = store.uiState;

	return (bdigg: BigNumber) => {
		if (!stats.stats.digg || !rebaseStats.btcPrice) return new BigNumber('0');
		const rebasePercentage = ((stats.stats.digg - rebaseStats.btcPrice) / rebaseStats.btcPrice) * 0.1;
		return bdigg.plus(bdigg.multipliedBy(rebasePercentage));
	};
}

