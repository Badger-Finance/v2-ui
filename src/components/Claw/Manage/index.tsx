import React from 'react';
import { ethers } from 'ethers';
import { observer } from 'mobx-react-lite';
import { Box, Container, Grid, MenuItem, Select } from '@material-ui/core';
import BigNumber from 'bignumber.js';
import { StoreContext } from 'mobx/store-context';

import { TokenSelectorLabel } from 'components-v2/common/TokenSelectorLabel';
import { TokenSelectorWithAmountContainer } from 'components-v2/common/TokenSelectorWithAmountContainer';
import { ActionButton } from '../ActionButton';
import { ClawDetails } from '../ClawDetails';
import { scaleToString, Direction, validateAmountBoundaries } from 'utils/componentHelpers';
import { useDetails, useError } from './manage.hooks';
import { ClawParam } from '../claw.model';
import { TokenSelect } from '../../../components-v2/common/TokenSelect';
import { TokenAmountInput } from '../../../components-v2/common/TokenAmountInput';
import { PercentageGroup } from '../../../components-v2/common/PercentageGroup';

enum Mode {
	DEPOSIT = 'deposit',
	WITHDRAW = 'withdraw',
}

const Manage = observer(() => {
	const { claw: store, contracts, wallet } = React.useContext(StoreContext);
	const { collaterals, claws, syntheticsDataByEMP, sponsorInformationByEMP } = store;
	const [mode, setMode] = React.useState<Mode.DEPOSIT | Mode.WITHDRAW>(Mode.DEPOSIT);
	const [manage, setManageParams] = React.useState<ClawParam>({});
	const details = useDetails(mode, manage);
	const error = useError(manage);

	const { selectedOption, amount } = manage;
	const selectedSynthetic = syntheticsDataByEMP.get(selectedOption || '');
	const bToken = contracts.tokens[selectedSynthetic?.collateralCurrency ?? ''];
	const decimals = bToken ? bToken.decimals : 18; // Default to 18 decimals.
	const position = sponsorInformationByEMP.get(selectedOption || '')?.position;

	// handleManage depends on the mode.
	let balanceLabel = '';
	let balance = new BigNumber(0);

	switch (mode) {
		case Mode.DEPOSIT:
			balance = bToken?.balance;
			balanceLabel = bToken && `Available ${collaterals.get(bToken.address)}: `;
			break;
		case Mode.WITHDRAW:
			balanceLabel = bToken && `Deposited collateral ${collaterals.get(bToken.address)}: `;
			if (position) {
				balance = position.rawCollateral;
			}
			break;
	}

	const handleManageFns = {
		[Mode.DEPOSIT]: async () => {
			const [empAddress, depositAmount] = [selectedOption, amount];
			if (!empAddress || !depositAmount) return;
			await store.actionStore.deposit(empAddress, ethers.utils.parseUnits(depositAmount, decimals).toHexString());
		},
		[Mode.WITHDRAW]: async () => {
			const [empAddress, collateralAmount] = [selectedOption, amount];
			if (!empAddress || !collateralAmount) return;
			await store.actionStore.withdraw(
				empAddress,
				ethers.utils.parseUnits(collateralAmount, decimals).toHexString(),
			);
		},
	};

	return (
		<Container>
			<Box pb={1}>
				<Grid item xs={12} sm={4} style={{ margin: 'auto' }}>
					<Select
						displayEmpty
						fullWidth
						variant="outlined"
						color="secondary"
						value={mode}
						style={{ textAlign: 'center' }}
						onChange={(v: any) => {
							setManageParams({});
							setMode(v.target.value);
						}}
					>
						<MenuItem value="" disabled>
							Select Mode
						</MenuItem>
						<MenuItem value={Mode.DEPOSIT}>DEPOSIT</MenuItem>
						<MenuItem value={Mode.WITHDRAW}>WITHDRAW</MenuItem>
					</Select>
				</Grid>
			</Box>
			<Grid item xs={12}>
				<Grid item xs={12}>
					<TokenSelectorWithAmountContainer
						tokenBalanceInformation={
							<TokenSelectorLabel
								name="Token"
								balanceLabel={balanceLabel}
								balance={selectedOption && scaleToString(balance, decimals, Direction.Down)}
							/>
						}
						tokenList={
							<TokenSelect
								selectedOption={selectedOption}
								options={claws}
								placeholder="Select Token"
								disabled={!wallet.connectedAddress || claws.size === 0}
								onChange={(selectedOption: string) => setManageParams({ selectedOption })}
							/>
						}
						tokenAmount={
							<TokenAmountInput
								value={amount}
								disabled={!selectedOption || !wallet.connectedAddress}
								onChange={(amount: string) => {
									if (!balance) return;
									setManageParams({
										selectedOption,
										amount,
										error: validateAmountBoundaries({
											amount,
											maximum: balance,
										}),
									});
								}}
							/>
						}
						percentagesGroup={
							<PercentageGroup
								disabled={!selectedOption || !wallet.connectedAddress}
								options={[25, 50, 75, 100]}
								onChange={(percentage: number) => {
									if (!balance) return;
									setManageParams({
										selectedOption,
										amount: balance
											.multipliedBy(percentage / 100)
											.dividedBy(10 ** decimals)
											.toFixed(decimals, BigNumber.ROUND_DOWN),
										error: undefined,
									});
								}}
							/>
						}
					/>
				</Grid>
				<Grid item xs={12}>
					<ClawDetails details={details} />
				</Grid>
				<Grid item xs={12}>
					<Grid container>
						<ActionButton
							text={error ? error : mode.toLocaleUpperCase()}
							disabled={!!error}
							onClick={handleManageFns[mode]}
						/>
					</Grid>
				</Grid>
			</Grid>
		</Container>
	);
});

export default Manage;
