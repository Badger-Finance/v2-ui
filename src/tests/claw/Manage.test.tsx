import React from 'react';
import '@testing-library/jest-dom';
import store from '../../mobx/store';
import {
	mockClaws,
	mockCollaterals,
	mockSponsorInformationByEmp,
	mockSyntheticDataByEmp,
	mockTokens,
} from './claw.mocks';
import { customRender, screen, fireEvent, within } from '../Utils';
import { StoreProvider } from '../../mobx/store-context';
import Manage from '../../components/Claw/Manage';
import * as ManageHooks from '../../components/Claw/Manage/manage.hooks';

describe('Claw - Manage', () => {
	beforeAll(() => {
		store.wallet.connectedAddress = '0x1a1a1a1a1a1a1a1a1a1a1a1a1a1a1a1a1a1a1a1a';
		store.contracts.tokens = mockTokens;
		store.claw.collaterals = mockCollaterals;
		store.claw.syntheticsDataByEMP = mockSyntheticDataByEmp;
		store.claw.sponsorInformationByEMP = mockSponsorInformationByEmp;
		store.claw.claws = mockClaws;
	});

	beforeEach(() => {
		customRender(
			<StoreProvider value={store}>
				<Manage />
			</StoreProvider>,
		);
	});

	test('matches snapshot', () => {
		const { container } = customRender(
			<StoreProvider value={store}>
				<Manage />
			</StoreProvider>,
		);
		expect(container).toMatchSnapshot();
	});

	test('mode select is displayed', () => {
		expect(screen.getByRole('button', { name: 'DEPOSIT' })).toBeInTheDocument();
	});

	test('mode can be changed', () => {
		fireEvent.mouseDown(screen.getByRole('button', { name: 'DEPOSIT' }));
		fireEvent.click(within(screen.getByRole('listbox')).getByText('WITHDRAW'));
		expect(screen.getByRole('button', { name: 'WITHDRAW' })).toBeInTheDocument();
	});

	test('token list has synthetics information', () => {
		const tokenSelector = screen.getByRole('button', { name: 'Select Token' });
		expect(tokenSelector).toBeEnabled();
		fireEvent.mouseDown(tokenSelector);
		const tokenList = within(screen.getByRole('listbox'));
		expect(tokenList.getByText('USD/bBadger 5-29')).toBeInTheDocument();
		expect(tokenList.getByText('USD-[bwBTC/ETH SLP] 5-29')).toBeInTheDocument();
	});

	describe('when no synthetic token is selected', () => {
		test('percentage group is disabled', () => {
			expect(
				screen.getByRole('button', {
					name: '50%',
				}),
			).toBeDisabled();
		});

		test('action button is disabled and with text "Select a Token"', () => {
			expect(screen.getByRole('button', { name: 'Select a Token' })).toBeDisabled();
		});
	});

	describe('when a synthetic token is selected', () => {
		beforeEach(() => {
			fireEvent.mouseDown(screen.getByRole('button', { name: 'Select Token' }));
			fireEvent.click(within(screen.getByRole('listbox')).getByText('USD/bBadger 5-29'));
		});

		test('its collateral token balance is displayed', () => {
			expect(screen.getByText('Available bBADGER:')).toBeInTheDocument();
			expect(screen.getByText('522.378430153821121982')).toBeInTheDocument();
		});

		test('action button is disabled and with text "Enter an amount"', () => {
			expect(screen.getByRole('button', { name: 'Enter an amount' })).toBeDisabled();
		});

		test('synthetic amount can be changed', () => {
			fireEvent.change(screen.getByRole('textbox'), { target: { value: '123' } });
			expect(screen.getByDisplayValue('123')).toBeInTheDocument();
		});

		test('collateral token balance percentage can applied', () => {
			fireEvent.click(
				screen.getByRole('button', {
					name: '50%',
				}),
			);
			expect(screen.getByDisplayValue('261.189215076910560991'));
		});
	});

	describe('Deposit Mode', () => {
		describe('when synthetic amount is inputted ', () => {
			beforeEach(() => {
				jest.clearAllMocks();
				jest.spyOn(ManageHooks, 'useDetails').mockReturnValue([
					{
						name: 'Liquidation Price',
						text: '6605.341175597608242402 $',
					},
					{
						name: 'Minimum Deposit',
						text: '100 CLAW',
					},
					{
						name: 'Collateral Ratio - Global',
						text: '0.001228214274254933 bBADGER',
					},
					{
						name: 'Collateral Ratio - Minimum',
						text: '1.2 $',
					},
					{
						name: 'Collateral Ratio - Current',
						text: '53.60346523838656428672 $',
					},
					{
						name: 'Expiration',
						text: 'May 29, 2021 22:00 UTC',
					},
				]);
				fireEvent.mouseDown(screen.getByRole('button', { name: 'Select Token' }));
				fireEvent.click(within(screen.getByRole('listbox')).getByText('USD/bBadger 5-29'));
				fireEvent.change(screen.getByRole('textbox'), { target: { value: '123' } });
			});

			test('action button is enabled', () => {
				const [, actionButton] = screen.getAllByRole('button', { name: 'DEPOSIT' });
				expect(actionButton).toBeEnabled();
			});

			test('deposit details are displayed', () => {
				expect(screen.getByText('6605.341175597608242402 $')).toBeInTheDocument();
				expect(screen.getByText('100 CLAW')).toBeInTheDocument();
				expect(screen.getByText('0.001228214274254933 bBADGER')).toBeInTheDocument();
				expect(screen.getByText('1.2 $')).toBeInTheDocument();
				expect(screen.getByText('53.60346523838656428672 $')).toBeInTheDocument();
				expect(screen.getByText('May 29, 2021 22:00 UTC')).toBeInTheDocument();
			});
		});
	});

	describe('Withdraw Mode', () => {
		describe('when synthetic amount is inputted ', () => {
			beforeEach(() => {
				jest.clearAllMocks();
				jest.spyOn(ManageHooks, 'useDetails').mockReturnValue([
					{
						name: 'Withdraw Speed',
						text: 'Slow',
					},
					{
						name: 'Minimum Deposit',
						text: '100 CLAW',
					},
					{
						name: 'Collateral Ratio - Global',
						text: '2.065952e-12 bSLP',
					},
					{
						name: 'Collateral Ratio - Minimum',
						text: '1.2 $',
					},
					{
						name: 'Collateral Ratio - Current',
						text: '54723969744.51006123772 $',
					},
					{
						name: 'Expiration',
						text: 'May 29, 2021 22:00 UTC',
					},
				]);
				const [modeSelect] = screen.getAllByRole('button', { name: 'DEPOSIT' });
				fireEvent.mouseDown(modeSelect);
				fireEvent.click(within(screen.getByRole('listbox')).getByText('WITHDRAW'));
				fireEvent.mouseDown(screen.getByRole('button', { name: 'Select Token' }));
				fireEvent.click(within(screen.getByRole('listbox')).getByText('USD-[bwBTC/ETH SLP] 5-29'));
				fireEvent.change(screen.getByRole('textbox'), { target: { value: '123' } });
			});

			test('action button is enabled', () => {
				const [, actionButton] = screen.getAllByRole('button', { name: 'WITHDRAW' });
				expect(actionButton).toBeEnabled();
			});

			test('deposit details are displayed', () => {
				expect(screen.getByText('Slow')).toBeInTheDocument();
				expect(screen.getByText('100 CLAW')).toBeInTheDocument();
				expect(screen.getByText('2.065952e-12 bSLP')).toBeInTheDocument();
				expect(screen.getByText('1.2 $')).toBeInTheDocument();
				expect(screen.getByText('54723969744.51006123772 $')).toBeInTheDocument();
				expect(screen.getByText('May 29, 2021 22:00 UTC')).toBeInTheDocument();
			});
		});
	});
});
