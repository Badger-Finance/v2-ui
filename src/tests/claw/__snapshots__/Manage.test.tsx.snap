// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Claw - Manage matches snapshot 1`] = `
<div>
  <div
    class="MuiContainer-root-7 MuiContainer-maxWidthLg-13"
  >
    <div
      class="MuiBox-root-15 MuiBox-root-1"
    >
      <div
        class="MuiGrid-root-17 MuiGrid-item-19 MuiGrid-grid-xs-12-63 MuiGrid-grid-sm-4-69"
        style="margin: auto;"
      >
        <div
          class="MuiInputBase-root-146 MuiOutlinedInput-root-131 MuiInputBase-colorSecondary-155 MuiOutlinedInput-colorSecondary-132 MuiInputBase-fullWidth-156"
          style="text-align: center;"
        >
          <div
            aria-haspopup="listbox"
            class="MuiSelect-root-120 MuiSelect-select-121 MuiSelect-selectMenu-124 MuiSelect-outlined-123 MuiInputBase-input-157 MuiOutlinedInput-input-141"
            role="button"
            tabindex="0"
          >
            DEPOSIT
          </div>
          <input
            aria-hidden="true"
            class="MuiSelect-nativeInput-130"
            tabindex="-1"
            value="deposit"
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root-164 MuiSelect-icon-126 MuiSelect-iconOutlined-129"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root-177 MuiOutlinedInput-notchedOutline-140"
            style="padding-left: 8px;"
          >
            <legend
              class="PrivateNotchedOutline-legend-178"
              style="width: 0.01px;"
            >
              <span>
                â€‹
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root-17 MuiGrid-item-19 MuiGrid-grid-xs-12-63"
    >
      <div
        class="MuiGrid-root-17 MuiGrid-item-19 MuiGrid-grid-xs-12-63"
      >
        <div
          class="MuiGrid-root-17 MuiGrid-container-18"
        >
          <div
            class="MuiGrid-root-17 MuiBox-root-15 MuiBox-root-2 MuiGrid-item-19 MuiGrid-grid-xs-12-63"
          >
            <div
              class="MuiGrid-root-17 MuiGrid-container-18 MuiGrid-spacing-xs-1-40"
            >
              <div
                class="MuiGrid-root-17 MuiGrid-item-19 MuiGrid-grid-xs-12-63 MuiGrid-grid-sm-6-71"
              >
                <p
                  class="MuiTypography-root-186 MuiTypography-body2-187 MuiTypography-colorTextSecondary-212"
                >
                  Token
                </p>
              </div>
              <div
                class="MuiGrid-root-17 MuiGrid-item-19 MuiGrid-grid-xs-12-63 MuiGrid-grid-sm-6-71"
              >
                <div
                  class="MuiGrid-root-17 makeStyles-availableText-184 MuiGrid-container-18 MuiGrid-spacing-xs-1-40"
                >
                  <div
                    class="MuiGrid-root-17 MuiGrid-item-19"
                  >
                    <p
                      class="MuiTypography-root-186 MuiTypography-body2-187 MuiTypography-colorTextSecondary-212"
                    />
                  </div>
                  <div
                    class="MuiGrid-root-17 makeStyles-balanceText-185 MuiGrid-item-19"
                  >
                    <p
                      class="MuiTypography-root-186 MuiTypography-body2-187 MuiTypography-colorTextPrimary-211"
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="MuiGrid-root-17 MuiGrid-item-19 MuiGrid-grid-xs-12-63"
          >
            <div
              class="MuiGrid-root-17 makeStyles-border-181 MuiBox-root-15 MuiBox-root-3 MuiGrid-container-18 MuiGrid-align-items-xs-center-26 MuiGrid-align-content-xs-center-30"
            >
              <div
                class="MuiGrid-root-17 MuiGrid-item-19 MuiGrid-grid-xs-12-63 MuiGrid-grid-sm-8-73 MuiGrid-grid-lg-9-102"
              >
                <div
                  class="MuiGrid-root-17 makeStyles-selectContainer-182 MuiGrid-container-18 MuiGrid-spacing-xs-2-41 MuiGrid-align-items-xs-center-26"
                >
                  <div
                    class="MuiGrid-root-17 MuiGrid-item-19 MuiGrid-grid-xs-12-63 MuiGrid-grid-sm-true-65"
                  >
                    <div
                      class="MuiFormControl-root-218 makeStyles-margin-217"
                    >
                      <div
                        class="MuiInputBase-root-146 MuiInputBase-formControl-147"
                      >
                        <div
                          aria-haspopup="listbox"
                          class="MuiSelect-root-120 MuiSelect-select-121 MuiSelect-selectMenu-124 MuiInputBase-input-157"
                          role="button"
                          tabindex="0"
                        >
                          Select Token
                        </div>
                        <input
                          aria-hidden="true"
                          class="MuiSelect-nativeInput-130"
                          placeholder="Select Token"
                          tabindex="-1"
                          value=""
                        />
                        <svg
                          aria-hidden="true"
                          class="MuiSvgIcon-root-164 MuiSelect-icon-126"
                          focusable="false"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M7 10l5 5 5-5z"
                          />
                        </svg>
                      </div>
                    </div>
                  </div>
                  <div
                    class="MuiGrid-root-17 MuiGrid-item-19 MuiGrid-grid-xs-12-63 MuiGrid-grid-sm-true-65"
                  >
                    <div
                      class="MuiInputBase-root-146 makeStyles-input-222 MuiInputBase-disabled-149 MuiInputBase-fullWidth-156"
                    >
                      <input
                        class="MuiInputBase-input-157 MuiInputBase-disabled-149"
                        disabled=""
                        pattern="^[0-9]*[.,]?[0-9]*$"
                        placeholder="0.00"
                        type="tel"
                        value=""
                      />
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="MuiGrid-root-17 MuiGrid-item-19 MuiGrid-grid-xs-12-63 MuiGrid-grid-sm-4-69 MuiGrid-grid-lg-3-96"
              >
                <div
                  class="MuiGrid-root-17 MuiGrid-container-18 MuiGrid-justify-xs-flex-end-36"
                >
                  <div
                    aria-label="text button group"
                    class="MuiButtonGroup-root-224"
                    role="group"
                  >
                    <button
                      class="MuiButtonBase-root-277 MuiButton-root-248 MuiButton-text-250 MuiButtonGroup-grouped-230 MuiButtonGroup-groupedHorizontal-231 MuiButtonGroup-groupedText-233 MuiButtonGroup-groupedTextHorizontal-234 MuiButtonGroup-groupedText-233 MuiButtonGroup-disabled-227 makeStyles-button-223 MuiButton-textSizeSmall-263 MuiButton-sizeSmall-269 MuiButton-disableElevation-259 MuiButton-disabled-261 MuiButtonBase-disabled-278"
                      disabled=""
                      tabindex="-1"
                      type="button"
                    >
                      <span
                        class="MuiButton-label-249"
                      >
                        25
                        %
                      </span>
                    </button>
                    <button
                      class="MuiButtonBase-root-277 MuiButton-root-248 MuiButton-text-250 MuiButtonGroup-grouped-230 MuiButtonGroup-groupedHorizontal-231 MuiButtonGroup-groupedText-233 MuiButtonGroup-groupedTextHorizontal-234 MuiButtonGroup-groupedText-233 MuiButtonGroup-disabled-227 makeStyles-button-223 MuiButton-textSizeSmall-263 MuiButton-sizeSmall-269 MuiButton-disableElevation-259 MuiButton-disabled-261 MuiButtonBase-disabled-278"
                      disabled=""
                      tabindex="-1"
                      type="button"
                    >
                      <span
                        class="MuiButton-label-249"
                      >
                        50
                        %
                      </span>
                    </button>
                    <button
                      class="MuiButtonBase-root-277 MuiButton-root-248 MuiButton-text-250 MuiButtonGroup-grouped-230 MuiButtonGroup-groupedHorizontal-231 MuiButtonGroup-groupedText-233 MuiButtonGroup-groupedTextHorizontal-234 MuiButtonGroup-groupedText-233 MuiButtonGroup-disabled-227 makeStyles-button-223 MuiButton-textSizeSmall-263 MuiButton-sizeSmall-269 MuiButton-disableElevation-259 MuiButton-disabled-261 MuiButtonBase-disabled-278"
                      disabled=""
                      tabindex="-1"
                      type="button"
                    >
                      <span
                        class="MuiButton-label-249"
                      >
                        75
                        %
                      </span>
                    </button>
                    <button
                      class="MuiButtonBase-root-277 MuiButton-root-248 MuiButton-text-250 MuiButtonGroup-grouped-230 MuiButtonGroup-groupedHorizontal-231 MuiButtonGroup-groupedText-233 MuiButtonGroup-groupedTextHorizontal-234 MuiButtonGroup-groupedText-233 MuiButtonGroup-disabled-227 makeStyles-button-223 MuiButton-textSizeSmall-263 MuiButton-sizeSmall-269 MuiButton-disableElevation-259 MuiButton-disabled-261 MuiButtonBase-disabled-278"
                      disabled=""
                      tabindex="-1"
                      type="button"
                    >
                      <span
                        class="MuiButton-label-249"
                      >
                        100
                        %
                      </span>
                    </button>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root-17 MuiGrid-item-19 MuiGrid-grid-xs-12-63"
      >
        <div
          class="MuiGrid-root-17 makeStyles-details-4 MuiGrid-container-18"
        >
          <div
            class="MuiGrid-root-17 makeStyles-root-280 MuiGrid-container-18"
          >
            <div
              class="MuiGrid-root-17 makeStyles-row-281 MuiGrid-container-18"
            >
              <div
                class="MuiGrid-root-17 makeStyles-detailName-282 MuiGrid-item-19 MuiGrid-grid-xs-6-57"
              >
                <p
                  class="MuiTypography-root-186 MuiTypography-body2-187 MuiTypography-colorTextSecondary-212"
                >
                  Liquidation Price
                </p>
              </div>
              <div
                class="MuiGrid-root-17 makeStyles-placeholder-284 MuiGrid-container-18 MuiGrid-item-19 MuiGrid-grid-xs-6-57"
              >
                <div
                  class="MuiGrid-root-17 MuiGrid-item-19 MuiGrid-grid-xs-12-63"
                >
                  <p
                    class="MuiTypography-root-186 MuiTypography-body2-187 MuiTypography-colorTextPrimary-211"
                  >
                    -
                  </p>
                </div>
              </div>
            </div>
            <div
              class="MuiGrid-root-17 makeStyles-row-281 MuiGrid-container-18"
            >
              <div
                class="MuiGrid-root-17 makeStyles-detailName-282 MuiGrid-item-19 MuiGrid-grid-xs-6-57"
              >
                <p
                  class="MuiTypography-root-186 MuiTypography-body2-187 MuiTypography-colorTextSecondary-212"
                >
                  Collateral Ratio - Global
                </p>
              </div>
              <div
                class="MuiGrid-root-17 makeStyles-placeholder-284 MuiGrid-container-18 MuiGrid-item-19 MuiGrid-grid-xs-6-57"
              >
                <div
                  class="MuiGrid-root-17 MuiGrid-item-19 MuiGrid-grid-xs-12-63"
                >
                  <p
                    class="MuiTypography-root-186 MuiTypography-body2-187 MuiTypography-colorTextPrimary-211"
                  >
                    -
                  </p>
                </div>
              </div>
            </div>
            <div
              class="MuiGrid-root-17 makeStyles-row-281 MuiGrid-container-18"
            >
              <div
                class="MuiGrid-root-17 makeStyles-detailName-282 MuiGrid-item-19 MuiGrid-grid-xs-6-57"
              >
                <p
                  class="MuiTypography-root-186 MuiTypography-body2-187 MuiTypography-colorTextSecondary-212"
                >
                  Collateral Ratio - Minimum
                </p>
              </div>
              <div
                class="MuiGrid-root-17 makeStyles-placeholder-284 MuiGrid-container-18 MuiGrid-item-19 MuiGrid-grid-xs-6-57"
              >
                <div
                  class="MuiGrid-root-17 MuiGrid-item-19 MuiGrid-grid-xs-12-63"
                >
                  <p
                    class="MuiTypography-root-186 MuiTypography-body2-187 MuiTypography-colorTextPrimary-211"
                  >
                    -
                  </p>
                </div>
              </div>
            </div>
            <div
              class="MuiGrid-root-17 makeStyles-row-281 MuiGrid-container-18"
            >
              <div
                class="MuiGrid-root-17 makeStyles-detailName-282 MuiGrid-item-19 MuiGrid-grid-xs-6-57"
              >
                <p
                  class="MuiTypography-root-186 MuiTypography-body2-187 MuiTypography-colorTextSecondary-212"
                >
                  Collateral Ratio - Current
                </p>
              </div>
              <div
                class="MuiGrid-root-17 makeStyles-placeholder-284 MuiGrid-container-18 MuiGrid-item-19 MuiGrid-grid-xs-6-57"
              >
                <div
                  class="MuiGrid-root-17 MuiGrid-item-19 MuiGrid-grid-xs-12-63"
                >
                  <p
                    class="MuiTypography-root-186 MuiTypography-body2-187 MuiTypography-colorTextPrimary-211"
                  >
                    -
                  </p>
                </div>
              </div>
            </div>
            <div
              class="MuiGrid-root-17 makeStyles-row-281 MuiGrid-container-18"
            >
              <div
                class="MuiGrid-root-17 makeStyles-detailName-282 MuiGrid-item-19 MuiGrid-grid-xs-6-57"
              >
                <p
                  class="MuiTypography-root-186 MuiTypography-body2-187 MuiTypography-colorTextSecondary-212"
                >
                  Expiration
                </p>
              </div>
              <div
                class="MuiGrid-root-17 makeStyles-placeholder-284 MuiGrid-container-18 MuiGrid-item-19 MuiGrid-grid-xs-6-57"
              >
                <div
                  class="MuiGrid-root-17 MuiGrid-item-19 MuiGrid-grid-xs-12-63"
                >
                  <p
                    class="MuiTypography-root-186 MuiTypography-body2-187 MuiTypography-colorTextPrimary-211"
                  >
                    -
                  </p>
                </div>
              </div>
            </div>
            <div
              class="MuiGrid-root-17 makeStyles-row-281 MuiGrid-container-18"
            >
              <div
                class="MuiGrid-root-17 makeStyles-detailName-282 MuiGrid-item-19 MuiGrid-grid-xs-6-57"
              >
                <p
                  class="MuiTypography-root-186 MuiTypography-body2-187 MuiTypography-colorTextSecondary-212"
                >
                  Minimum Deposit
                </p>
              </div>
              <div
                class="MuiGrid-root-17 makeStyles-placeholder-284 MuiGrid-container-18 MuiGrid-item-19 MuiGrid-grid-xs-6-57"
              >
                <div
                  class="MuiGrid-root-17 MuiGrid-item-19 MuiGrid-grid-xs-12-63"
                >
                  <p
                    class="MuiTypography-root-186 MuiTypography-body2-187 MuiTypography-colorTextPrimary-211"
                  >
                    -
                  </p>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root-17 MuiGrid-item-19 MuiGrid-grid-xs-12-63"
      >
        <div
          class="MuiGrid-root-17 MuiGrid-container-18"
        >
          <button
            class="MuiButtonBase-root-277 MuiButton-root-248 MuiButton-contained-256 makeStyles-button-285 MuiButton-containedPrimary-257 MuiButton-containedSizeLarge-268 MuiButton-sizeLarge-270 MuiButton-disabled-261 MuiButtonBase-disabled-278"
            disabled=""
            tabindex="-1"
            type="button"
          >
            <span
              class="MuiButton-label-249"
            >
              Select a Token
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
